main.go

```diff
@@ -2,16 +2,101 @@ package main
 
 import (
 	"fmt"
+	"log"
+	"os"
+	"runtime"
 	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/zapr"
+	"github.com/urfave/cli/v2"
+	"github.com/vshn/go-bootstrap/cfg"
+	"github.com/vshn/go-bootstrap/cmd"
+	"github.com/vshn/go-bootstrap/cmd/example"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
 )
 
 var (
 	// these variables are populated by Goreleaser when releasing
 	version = "unknown"
 	commit  = "-dirty-"
 	date    = time.Now().Format("2006-01-02")
+
+	// TODO: Adjust app name
+	appName     = "go-bootstrap"
+	appLongName = "a generic bootstrapping project"
 )
 
 func main() {
-	fmt.Printf("version %s, commit %s, date %s\n", version, commit, date)
+	err := app().Run(os.Args)
+	if err != nil {
+		log.Fatalf("unable to start %s: %v", appName, err)
+	}
+}
+
+func before(c *cli.Context) error {
+	logger := newLogger(appName, c.Bool("debug"))
+	cmd.SetAppLogger(c, logger)
+	if c.Args().Present() {
+		logger.WithValues(
+			"version", version,
+			"date", date,
+			"commit", commit,
+			"go_os", runtime.GOOS,
+			"go_arch", runtime.GOARCH,
+			"go_version", runtime.Version(),
+			"uid", os.Getuid(),
+			"gid", os.Getgid(),
+		).Info("Starting up " + appName)
+	}
+
+	return nil
+}
+
+func app() *cli.App {
+	cli.VersionPrinter = func(_ *cli.Context) {
+		fmt.Printf("version=%s revision=%s date=%s\n", version, commit, date)
+	}
+
+	compiled, err := time.Parse(time.RFC3339, date)
+	if err != nil {
+		compiled = time.Time{}
+	}
+
+	return &cli.App{
+		Name:     appName,
+		Usage:    appLongName,
+		Version:  version,
+		Compiled: compiled,
+
+		EnableBashCompletion: true,
+
+		Before: before,
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:    "debug",
+				Aliases: []string{"verbose", "d"},
+				Usage:   "sets the log level to debug",
+				EnvVars: []string{cfg.Env("DEBUG")},
+			},
+		},
+		Commands: []*cli.Command{
+			example.Command,
+		},
+	}
+}
+
+func newLogger(name string, debug bool) logr.Logger {
+	zc := zap.NewDevelopmentConfig()
+	if debug {
+		// Zap's levels get more verbose as the number gets smaller,
+		// bug logr's level increases with greater numbers.
+		zc.Level = zap.NewAtomicLevelAt(zapcore.Level(-2)) // max logger.V(2)
+	}
```

Comment:
Do you guys have strong opinions regarding the logging layout?
for example, `zap.NewProductionConfig()` formats everything to JSON. This makes it easier for external logging stacks like EFK, but harder to read for humans.

Would it make sense to automatically switch based on version number or or make it configurable?

Reply:
I prefer json. Especially in combination with EFK and Loki. I don't mind reading json but each to their own.

My preference would be:
- JSON is the default
- Add paramter `--log-format=pretty` or something for "prettier" logs.

Reply:
> Do you guys have strong opinions regarding the logging layout?

Yes, absolutely! I find it unreadable in the proposed format, it's the same in K8up, logs are barely readable by a human. It would be good to have a well-formated human-readable log-format, and I'm in favor of making this configurable.

Reply:
Do you have examples for "well-formated human-readable log-format"? I'm not sure what to imagine...

Reply:
I have to search for it, will report back.

Reply:
We could go fancy with colors like https://github.com/ccremer/plogr, but I'm not sure how well that works out in Kubernetes environments :smile: 
(it also introduces soft-dependency to pterm...)

Reply:
```
2022-01-06T12:22:56.976+0100 [INFO] go-bootstrap: Starting up go-bootstrap {"version": "unknown", "date": "2022-01-06", "commit": "-dirty-", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17.5", "uid": 1000, "gid": 1000}
2022-01-06T12:22:56.976+0100 [INFO] go-bootstrap.example: Hello from example command! {"config": {"ExampleFlag":"bar"}}
```

Maybe something like that, mainly getting rid of the tabs and only use spaces.

Reply:
Or:

```
level=info ts=2022-01-06T12:57:05.15609226Z caller=table_manager.go:363 msg="creating table" table=index_2526
level=info ts=2022-01-06T12:57:05.156099712Z caller=table_manager.go:363 msg="creating table" table=index_2553
level=info ts=2022-01-06T12:57:05.156107167Z caller=table_manager.go:363 msg="creating table" table=index_2571
level=info ts=2022-01-06T12:57:05.156114867Z caller=table_manager.go:363 msg="creating table" table=index_2652
level=info ts=2022-01-06T12:57:05.156122608Z caller=table_manager.go:363 msg="creating table" table=index_2632
level=info ts=2022-01-06T12:57:05.156130375Z caller=table_manager.go:363 msg="creating table" table=index_2538
level=info ts=2022-01-06T12:57:05.156138188Z caller=table_manager.go:363 msg="creating table" table=index_2558
level=info ts=2022-01-06T12:57:05.156147612Z caller=table_manager.go:363 msg="creating table" table=index_2574
level=info ts=2022-01-06T12:57:05.156155085Z caller=table_manager.go:363 msg="creating table" table=index_2593
level=info ts=2022-01-06T12:57:05.156162643Z caller=table_manager.go:363 msg="creating table" table=index_2616
```

Reply:
Or:

```
level=error ts=2021-12-29T10:41:18.697Z caller=notify.go:367 component=dispatcher msg="Error on notify" err="Post http://signalilo/webhook: context deadline exceeded"
level=error ts=2021-12-29T10:41:18.697Z caller=dispatch.go:264 component=dispatcher msg="Notify for alerts failed" num_alerts=4 err="Post http://signalilo/webhook: context deadline exceeded"
```

Very similar to above...

Reply:
Turns out with Zap I can't really customize the console format except for `ConsoleSeparator`
With `ConsoleSeparator = " "`
```console
2022-01-06T14:48:13.585+0100 INFO go-bootstrap Starting up go-bootstrap {"version": "unknown", "date": "2022-01-06", "commit": "-dirty-", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17.5", "uid": 1000, "gid": 1000}
2022-01-06T14:48:13.585+0100 INFO go-bootstrap.example Hello from example command! {"config": {"ExampleFlag":"bar"}}
```
With `ConsoleSeparator = ": "`
```console
2022-01-06T14:54:23.155+0100: INFO: go-bootstrap: Starting up go-bootstrap: {"version": "unknown", "date": "2022-01-06", "commit": "-dirty-", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17.5", "uid": 1000, "gid": 1000}
2022-01-06T14:54:23.155+0100: INFO: go-bootstrap.example: Hello from example command!: {"config": {"ExampleFlag":"bar"}}
```
With `ConsoleSeparator = " | "`
```console
2022-01-06T14:54:44.897+0100 | INFO | go-bootstrap | Starting up go-bootstrap | {"version": "unknown", "date": "2022-01-06", "commit": "-dirty-", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17.5", "uid": 1000, "gid": 1000}
2022-01-06T14:54:44.897+0100 | INFO | go-bootstrap.example | Hello from example command! | {"config": {"ExampleFlag":"bar"}}
```
If we want more customization, we'll need our own implementation of a console encoder (default is https://github.com/uber-go/zap/blob/master/zapcore/console_encoder.go)

Reply:
@tobru See my previous comment regarding the log format. Which format do you familiar best?
atm the moment it's
```console
2022-01-06T16:44:24.903+0100 | INFO | go-bootstrap | go-bootstrap/main.go:98 | Starting up go-bootstrap | {"version": "unknown", "date": "2022-01-06", "commit": "-dirty-", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17.5", "uid": 1000, "gid": 1000}
2022-01-06T16:44:24.903+0100 | DEBUG | go-bootstrap.example | go-bootstrap/example_command.go:31 | validating config
2022-01-06T16:44:24.903+0100 | INFO | go-bootstrap.example | go-bootstrap/example_command.go:40 | Hello from example command! | {"config": {"ExampleFlag":"asdf"}}
```

Reply:
I can live with that :+1: